src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+        return result }
+    let writeEvents (log : ILogger) retryPolicy (conn : MessageDbClient) (streamName : string) (version : int64) (events : IEventData<EventBody> array)
+        : Async<MdbSyncResult> =
+        let call = writeEventsLogged conn streamName version events
+        Log.withLoggedRetries retryPolicy "writeAttempt" call log
+
+module private Read =
+    let private toSlice (events : ITimelineEvent<EventBody> array) isLast =
+        let lastVersion = match Array.tryLast events with Some ev -> ev.Index | None -> -1L
+        { Messages = events; IsEnd = isLast; LastVersion = lastVersion }
+
+    let private readSliceAsync (conn : MessageDbClient) (streamName : string) (batchSize : int64) (startPos : int64) ct
+        : Task<_> = task {
+        let! page = conn.ReadStream(streamName, startPos, batchSize, ct)
+        let isLast = int64 page.Length < batchSize
+        return toSlice page isLast }
+
+    let private readLastEventAsync (conn : MessageDbClient) (streamName : string) ct : Task<_> = task {
+        let! events = conn.ReadLastEvent(streamName, ct)
+        return toSlice events false }
+
+    let inline len (bytes: EventBody) = bytes.Length
+    let resolvedEventLen (x : ITimelineEvent<EventBody>) = len x.Data + len x.Meta
+    let private loggedReadSlice conn streamName batchSize startPos (log : ILogger) : Async<_> = async {
+        let! ct = Async.CancellationToken
+        let! t, slice = readSliceAsync conn streamName batchSize startPos ct |> Async.AwaitTaskCorrect |> Stopwatch.Time
+        let bytes, count = slice.Messages |> Array.sumBy resolvedEventLen, slice.Messages.Length
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let evt = Log.Slice reqMetric
+        let log = if not (log.IsEnabled Events.LogEventLevel.Debug) then log else log |> Log.propResolvedEvents "Json" slice.Messages
+        (log |> Log.prop "startPos" startPos |> Log.prop "bytes" bytes |> Log.event evt).Information("Mdb{action:l} count={count} version={version}",
+            "Read", count, slice.LastVersion)
+        return slice }
+
+    let private readBatches (log : ILogger) (readSlice : int64 -> ILogger -> Async<StreamEventsSlice>)
+            (maxPermittedBatchReads : int option) (startPosition : int64)
+        : AsyncSeq<int64 * ITimelineEvent<EventBody> array> =
+        let rec loop batchCount pos : AsyncSeq<int64 * ITimelineEvent<EventBody> array> = asyncSeq {
+            match maxPermittedBatchReads with
+            | Some mpbr when batchCount >= mpbr -> log.Information "batch Limit exceeded"; invalidOp "batch Limit exceeded"
+            | _ -> ()
+
+            let batchLog = log |> Log.prop "batchIndex" batchCount
+            let! slice = readSlice pos batchLog
+
+            match slice.Messages with
+            | [||] -> yield -1L, Array.empty
+            | _ ->
```

Comment:
```suggestion
            | msgs ->
```
use on 210
(or replace with an `if`-`else` ...)

Reply:
Is this match even doing anything?

If it's an empty list then the version should be -1 and it should be the end. I think I can replace this entire thing with

```fsharp
yield slice.LastVersion, slice.Messages
if not slice.IsEnd then
    yield! loop (batchCount + 1) (slice.LastVersion + 1L) }
```

Reply:
it feels like you should be right; it's a long time since I've tried to reason about it. This is where we find out the whether the test code covers it ;)